### PR TITLE
TE-10424 Adjusted arch sniffer to check return type for Bridge methods

### PR DESCRIPTION
**Pull request by geega**
- Developer(s): @geega

- Ticket: https://spryker.atlassian.net/browse/TE-10424

- Core changes: https://github.com/spryker/spryker/pull/8999

- Release Group: https://release.spryker.com/release-groups/view/3927

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   ArchitectureSniffer   | patch |                       |

-----------------------------------------

##### Change log

Improvements

- Introduced `BridgeMethodsInterfaceRule` to check return type for bridge methods and interfaces, bridges and their interface should have return/param types added where possible.
- Adjusted `BridgeMethodsRule ` to check for method signature of bridge vs interface. They must always fully match.
- Adjusted ServiceRule so it skips the check for AbstractService.

### CODE REVIEW

**Comment by dereuromark** [comment]
Template (and changelog) is invalid in multiple places and needs fixing
This needs some tests as there are changes egde cases will fail under the current code, e.g. `...|array<type>` etc

**Comment by geega** [comment]
> Template (and changelog) is invalid in multiple places and needs fixing This needs some tests as there are changes egde cases will fail under the current code, e.g. `...|array<type>` etc

You about unit tests for different cases?

**Comment by dereuromark** [comment]
Always run validation in RG - and it will tell you why the template and release type is invalid for us.

File: src/Common/Bridge/BridgeMethodsInterfaceRule.php [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+            if (!$interfaceMethodReturnType) {
+                $errors[] = 'Missed return type.';
+            }
+
+            if ($interfaceMethodReturnType && $parentInterfaceReturnType && $interfaceMethodReturnType !== $parentInterfaceReturnType) {
+                $errors[] = 'Invalid return type.';
+            }
+
+            if (count($errors)) {
```
Comment (dereuromark) [inline src/Common/Bridge/BridgeMethodsInterfaceRule.php:137]: if ($errors) is sufficient.

File: src/Common/Bridge/BridgeMethodsInterfaceRule.php [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        }
+
+        $countParameters = count($birdgeInterfaceMethodParameters);
+
+        for ($i = 0; $i < $countParameters; $i++) {
+            $birdgeInterfaceMethodParameter = $birdgeInterfaceMethodParameters[$i];
+            $parentInterfaceMethodParameter = $parentInterfaceMethodParameters[$i];
+
+            if (!$birdgeInterfaceMethodParameter->getType() || !$parentInterfaceMethodParameter->getType()) {
+                $errors[] = '';
```
Comment (dereuromark) [inline src/Common/Bridge/BridgeMethodsInterfaceRule.php:171]: Why setting an empty string?

File: src/ArchitectureSnifferFactoryAwareTrait.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+/**
+ * Copyright © 2016-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace ArchitectureSniffer;
+
+trait ArchitectureSnifferFactoryAwareTrait
```
Comment (dereuromark) [inline src/ArchitectureSnifferFactoryAwareTrait.php:10]: Why do we need this?

File: src/ArchitectureSnifferFactory.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * Copyright © 2016-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace ArchitectureSniffer;
+
+use phpDocumentor\Reflection\DocBlock;
+use phpDocumentor\Reflection\DocBlockFactory;
+
+class ArchitectureSnifferFactory
```
Comment (dereuromark) [inline src/ArchitectureSnifferFactory.php:13]: Why do we need this?

Comment (geega) [reply src/ArchitectureSnifferFactory.php:13]: Yes. I will delete trait and factory files.

Comment (geega) [reply src/ArchitectureSnifferFactory.php]: Deleted files

File: src/Common/Bridge/BridgeMethodsInterfaceRule.php [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            }
+
+            $bridgedInterfaceReflectionMethod = $bridgedInterfaceReflection->getMethod($interfaceMethod->getName());
+
+            $interfaceMethodName = sprintf('%s::%s', $interfaceNode->getFullQualifiedName(), $interfaceMethod->getName());
+            $interfaceMethodReflection = new ReflectionMethod($interfaceMethodName);
+
+            $errors = $this->compareBridgeAndParentMethodInterface($interfaceMethodReflection, $bridgedInterfaceReflectionMethod);
+
+            $interfaceMethodReturnType = $this->reflactionReturnTypeToString($interfaceMethodReflection);
```
[the remark this answers is not in the document]
Comment (gechetspr) [reply src/Common/Bridge/BridgeMethodsInterfaceRule.php:126]: also method names should be verbs

File: src/Common/Bridge/BridgeMethodsInterfaceRule.php [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+            $parentInterfaceReturnType = $this->getReturnTypeFromParentInterface($interfaceMethod, $bridgedInterfaceReflection);
+
+            if (!$interfaceMethodReturnType) {
```
Comment (gechetspr) [inline src/Common/Bridge/BridgeMethodsInterfaceRule.php:129]: so we are forcing to add return types in bridge interfaces?

File: src/Common/Bridge/BridgeMethodsInterfaceRule.php [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+
+        return null;
+    }
+
+    /**
+     * @param array<\ReflectionParameter> $params
+     *
+     * @return int
+     */
+    protected function countRequireParams(array $params): int
```
Comment (gechetspr) [inline src/Common/Bridge/BridgeMethodsInterfaceRule.php:228]: required

**Review by gechetspr — CHANGES_REQUESTED** [review]

File: src/Common/Bridge/BridgeMethodsInterfaceRule.php (hunk 6 above)
Comment (dereuromark) [reply src/Common/Bridge/BridgeMethodsInterfaceRule.php:129]: We are at least making it visible now if they are missing

Comment (geega) [reply src/Common/Bridge/BridgeMethodsInterfaceRule.php:129]: Yes.

File: src/Common/Bridge/BridgeMethodsInterfaceRule.php [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+     * @param \PHPMD\Node\InterfaceNode $interfaceNode
+     *
+     * @return void
+     */
+    protected function verifyInterface(ClassNode $node, InterfaceNode $interfaceNode): void
+    {
+        $bridgedInterfaceReflection = $this->getBridgedInterfaceReflection($node->getMethods());
+        if ($bridgedInterfaceReflection === null) {
+            $message = sprintf(
+                'The bridge is missing an interface. That violates the rule "%s"',
```
[the remark this answers is not in the document]
Comment (geega) [reply src/Common/Bridge/BridgeMethodsInterfaceRule.php:78]: This error can happens when parent interface not found in doc block 
Like this `\Spryker\Service\UtilText\UtilTextServiceInterface`

```
    /**
     * @param \Spryker\Service\UtilText\UtilTextServiceInterface $utilTextService
     */
    public function __construct($utilTextService)
    {
        $this->utilTextService = $utilTextService;
    }
```

[the remark this answers is not in the document]
Comment (geega) [reply src/Common/Bridge/BridgeMethodsInterfaceRule.php:78]: I will change to 
`The bridge constructor doc block is missing a parent interface. `

File: src/Common/Bridge/BridgeMethodsInterfaceRule.php [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+            $bridgedInterfaceReflectionMethod = $bridgedInterfaceReflection->getMethod($interfaceMethod->getName());
+
+            $interfaceMethodName = sprintf('%s::%s', $interfaceNode->getFullQualifiedName(), $interfaceMethod->getName());
+
+            $interfaceMethodReflection = new ReflectionMethod($interfaceMethodName);
+
+            $errors = $this->compareBridgeAndParentMethodInterface($interfaceMethodReflection, $bridgedInterfaceReflectionMethod);
+
+            $interfaceMethodReturnType = $this->reflectionReturnTypeToString($interfaceMethodReflection);
```
Comment (gechetspr) [inline src/Common/Bridge/BridgeMethodsInterfaceRule.php:127]: still not a verb

**Review by gechetspr — APPROVED** [review]
Everything looks good except method name. Can be merged after the fix

**Comment by dereuromark** [comment]
@gechetspr 

Prio 1
- FacadeReturnValueRule: 32
- LayerAccessRule: 15
- FactoryNoLoopsRule: 11
- FactoryCreateContainOneNewRule: 10
- FactoryOnlyGetAndCreateRule: 15
- FactoryGetContainNoNewRule: 2
- FacadeRule: 9
- FacadeNoLogicRule: 7
- AllMethodsPublicInFacadeRule: 3
- ClientInterfaceRule: 1
Prio 2
- BridgeMethodsInterfaceRule: 2212
- BridgeMethodsRule: 18

I wonder if we can fix some of those issues soonish.

**Pull request merged**